Update repo-config from v0.11.0 to v0.13.0

camille-bouvy-frequenz opened the pull request.

Commit:
Update repo-config from v0.11.0 to v0.13.0

Signed-off-by: camille-bouvy-frequenz <camille.bouvy@frequenz.com>

Commit:
Update release notes

Signed-off-by: camille-bouvy-frequenz <camille.bouvy@frequenz.com>

Commit:
Apply the repo-config v0.12.0 migration script

Signed-off-by: camille-bouvy-frequenz <camille.bouvy@frequenz.com>

ghost commented:
Hey @shsms, I recon adding the following in `pyproject.toml`:
```
[tool.pytest.ini_options]
+addopts = "-W=all -Werror -Wdefault::DeprecationWarning -Wdefault::PendingDeprecationWarning -vv"
```
now breaks the `nox` tests. It comes from the base client, should we fix it there and create a new release?

shsms commented:
It should be fine to create an issue in the base client and then exclude the UnraiseableException from the list.

llucax commented:
I just had a look at the failing tests and damn, things got quite complicated in those tests. Is there any reason why you are manually creating the async loop instead of using pytest_asyncio and configure it to have a function scope?

Also the mocking of async iterators seems to be pretty messed up/wrong. Streaming functions are mocked but not the iterator they return.

It could definitely take some time to fix those tests... :-/

llucax commented:
I pushed a commit to ignore warnings in the problematic tests for now, I leave creating the issue to you @shsms

llucax committed:
Avoid grouping a bunch of dependencies still at v0.x

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
OK, tests still fail in the merge queue, it seems to be very flaky and affect most tests.

I'm having a second look and think I can fix the warning for real.

llucax committed:
Mock async iterators properly

By default AsyncMock doesn't provide any proper mocking of async
async iterators, and if we just assign a `AsyncMock` to stub calls like
`ReceivePublicTradesStream`, what happens is that
`async for x in ReceivePublicTradesStream()` is used instead. What
Python does with this is something like:

```py
stream = ReceivePublicTradesStream()
it = stream.__aiter__()  # Gets an anync iterator, but this call is sync
while v := await it.__anext():
    ...
```

Because `ReceivePublicTradesStream` is an `AsyncMock`, getting any
attributes (like `__aiter__()`) results in returning a new `AsyncMock`.

But since `__aiter__()` is called without an `await`, then we get the
following exceptions/warnings:

* TypeError("'async for' requires an object with __aiter__ method, got
  coroutine")
* RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was
  never awaited

To fix this, we just crate a proper fake async iterator we can return
to the async mock.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Re-format pyproject.toml using taplo

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Close the loop on every async test function

This helps detecting issues with cleanup in the place where they happen.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove unnecessary @pytest.mark.asyncio

We are already using the option to auto-use the `pytest-asyncio` plugin
for every `async` function, so we don't need to mark them manually.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Don't use a custom asyncio loop in test

There is really no need to create our own loop for these tests, we can
just use the loop provided by `pytest-asyncio` which already offers a
lot of tools for detecting issues and configure how we want the loop to
work.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
OK, first approach didn't work out. There were more issues. I think I have all of them fixed now, it took a bit longer than what I would have wanted, but I hope we can continue with a warning-free project now :crossed_fingers: 

If tests pass I will remove the commit ignoring the warnings and the revert.

It was really tricky to find out what was going on because `pytest-asyncio` wasn't configured to create one loop per test function, so cleanup issues in one test would appear in a completely different test, even in a different file. :unamused:

llucax commented:
Test seem to be passing now, will do the final rebasing and activate auto-merge.

llucax commented:
BTW, ready to merge again, it just needs a new approval @matthias-wende-frequenz
